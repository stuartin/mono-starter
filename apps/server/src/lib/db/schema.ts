import { createId } from "@paralleldrive/cuid2";
import { integer, sqliteTable, text } from "drizzle-orm/sqlite-core";
import { createSchemaFactory } from "drizzle-zod";

export const { createInsertSchema, createSelectSchema, createUpdateSchema } = createSchemaFactory({
    coerce: {
        date: true,
    },
});


export const user_table = sqliteTable("user", {
    id: text().primaryKey().$defaultFn(() => createId()),
    name: text().notNull(),
    email: text().notNull(),
    passwordHash: text().notNull(),
    updatedAt: integer({ mode: "timestamp" })
        .notNull()
        .$defaultFn(() => new Date())
        .$onUpdateFn(() => new Date()),
    createdAt: integer({ mode: "timestamp" })
        .notNull()
        .$defaultFn(() => new Date()),
});

export const session_table = sqliteTable("session", {
    id: text().primaryKey(),
    userId: integer().notNull().references(() => user_table.id),
    expiresAt: integer({ mode: "timestamp" }).notNull()
});