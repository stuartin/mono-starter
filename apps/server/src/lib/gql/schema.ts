import { builder } from "./builder";

import "./areas";
import "./climbs";
import "./geo";
import "./grades";
import "./guides";
import "./medias";
import "./notices";
import "./systems";
import "./topos";

export const schema = builder.toSchema();
