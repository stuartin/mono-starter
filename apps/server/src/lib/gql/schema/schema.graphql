schema {
  query: Query
  mutation: Mutation
}

"""Area"""
type Area {
  climbs: [Climb!]
  createdAt: DateTime
  guide: Guide
  id: ID
  name: String
  parentArea: Area
  rockTypes: [AreaToRockType!]
  slug: String
  topos: [Topo!]
  updatedAt: DateTime
}

type AreaToRockType {
  area: Area
  rockType: RockType
}

"""Climb"""
type Climb {
  area: Area
  climbType: ClimbType
  createdAt: DateTime
  description: String
  fa: String
  gear: String
  grade: Grade
  guide: Guide
  height: Int
  id: ID
  name: String
  pitches: [Pitch!]
  slug: String
  updatedAt: DateTime
}

"""ClimbType"""
type ClimbType {
  createdAt: DateTime
  id: ID
  name: ClimbTypeNameRef
  updatedAt: DateTime
}

enum ClimbTypeNameRef {
  boulder
  mixed
  sport
  trad
}

input CountriesFilterInput {
  id: ID
  name: String
}

"""Country"""
type Country {
  createdAt: DateTime
  id: ID
  name: String
  states: [State!]
  updatedAt: DateTime
}

"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

"""Grade"""
type Grade {
  createdAt: DateTime
  gradeSystem: GradeSystem
  id: ID
  maxScore: Int
  minScore: Int
  updatedAt: DateTime
  value: String
}

"""GradeSystem"""
type GradeSystem {
  createdAt: DateTime
  id: ID
  name: String
  shortName: String
  updatedAt: DateTime
}

"""Guide"""
type Guide {
  areas: [Area!]
  country: Country
  createdAt: DateTime
  gradeSystems: [GuideToGradeSystem!]
  heightSystem: HeightSystem
  id: ID
  name: String
  notices: [Notice!]
  slug: String
  state: State
  updatedAt: DateTime
  visibility: VisibilityRef
}

input GuideFilterInput {
  id: ID
  name: String
}

type GuideToGradeSystem {
  gradeSystem: GradeSystem
  guide: Guide
}

"""HeightSystem"""
type HeightSystem {
  createdAt: DateTime
  id: ID
  name: HeightSystemNameRef
  updatedAt: DateTime
}

enum HeightSystemNameRef {
  ft
  m
}

"""Media"""
type Media {
  alt: String
  area: Area
  climb: Climb
  createdAt: DateTime
  guide: Guide
  height: Int
  id: ID
  updatedAt: DateTime
  url: String
  width: Int
}

type Mutation {
  addGuide(countryId: ID!, description: String!, gradeSystemIds: [ID!]!, heightSystemId: ID!, name: String!, stateId: ID): Guide
  deleteGuide(id: ID!): Boolean
  updateGuide(countryId: String, description: String, gradeSystemIds: [String!], heightSystemId: String, id: ID!, name: String, stateId: String): Guide
}

"""Notice"""
type Notice {
  area: Area
  climb: Climb
  createdAt: DateTime
  description: String
  guide: Guide
  id: ID
  summary: String
  updatedAt: DateTime
}

"""Pitch"""
type Pitch {
  climb: Climb
  createdAt: DateTime
  description: String
  grade: Grade
  id: ID
  updatedAt: DateTime
}

type Query {
  countries(filter: CountriesFilterInput, limit: Int! = 10): [Country!]
  country(filter: CountriesFilterInput!): Country
  guide(filter: GuideFilterInput!): Guide
  guides(filter: GuideFilterInput): [Guide!]
}

"""RockType"""
type RockType {
  createdAt: DateTime
  id: ID
  name: String
  updatedAt: DateTime
}

"""State"""
type State {
  createdAt: DateTime
  id: ID
  name: String
  updatedAt: DateTime
}

"""Topo"""
type Topo {
  area: Area
  createdAt: DateTime
  guide: Guide
  id: ID
  media: Media
  updatedAt: DateTime
}

enum VisibilityRef {
  private
  public
}