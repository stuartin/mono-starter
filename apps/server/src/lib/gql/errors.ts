export type ErrorExtensions = (typeof ERR)[keyof typeof ERR]["extensions"];

export const ERR = {
    NOT_FOUND: {
        extensions: {
            code: "NOT_FOUND",
        },
    },
    INVALID: {
        extensions: {
            code: "INVALID",
        },
    },
    SERVER: {
        extensions: {
            code: "SERVER",
            http: { status: 500 },
        },
    },
    UNAUTHORIZED: {
        extensions: {
            code: "UNAUTHORIZED",
            http: { status: 401 },
        },
    },
} as const;
