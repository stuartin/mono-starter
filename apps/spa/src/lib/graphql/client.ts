import { PUBLIC_SERVER_ENDPOINT } from '$env/static/public';
import { Client, debugExchange, fetchExchange, cacheExchange as documentCache } from '@urql/svelte';
import { cacheExchange } from "@urql/exchange-graphcache"
import { devtoolsExchange } from '@urql/devtools';

export const client = new Client({
    url: PUBLIC_SERVER_ENDPOINT,
    // ORDER MATTERS
    // exchanges: [devtoolsExchange, debugExchange, documentCache, fetchExchange],
    exchanges: [
        devtoolsExchange,
        debugExchange,
        cacheExchange(),
        fetchExchange
    ],
});