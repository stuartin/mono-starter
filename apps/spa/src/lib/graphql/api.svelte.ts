import { client } from "$lib/graphql/client";
import { QueryMe } from "$lib/graphql/queries";
import { queryStore, } from "@urql/svelte";
import { setContext, getContext } from "svelte";

export class Queries {

    queryMe = queryStore({
        client,
        query: QueryMe,
        context: { requestPolicy: 'cache-and-network' }
    })

}

const KEY = Symbol("API")

export function setAPI() {
    return setContext(KEY, new Queries())
}

export function getAPI() {
    return getContext<ReturnType<typeof setAPI>>(KEY)
}