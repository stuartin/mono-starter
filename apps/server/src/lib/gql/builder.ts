import type { Context } from "#lib/yoga";

import SchemaBuilder from "@pothos/core";
import DrizzlePlugin from "@pothos/plugin-drizzle";
import PothosScopeAuthPlugin from "@pothos/plugin-scope-auth";
import ZodValidationPlugin from "@pothos/plugin-zod";
import WithInputPlugin from '@pothos/plugin-with-input';
import SimpleObjectsPlugin from '@pothos/plugin-simple-objects';

import * as schema from "#lib/db/schema";
import { GraphQLError } from "graphql";
import { DateTimeResolver } from "graphql-scalars";

import { ERR } from "./errors";
import { drizzle } from "#lib/db";
import { ValidAuthContext } from "#features/auth/context";

interface Root<T> {
    Context: T;
    DefaultInputFieldRequiredness: true;
    DrizzleSchema: typeof schema;
    AuthScopes: {
        user: boolean;
    };
    AuthContexts: {
        user: T & ValidAuthContext;
    };
    Scalars: {
        DateTime: {
            Output: Date;
            Input: Date;
        };
    };
}

export const builder = new SchemaBuilder<Root<Context>>({
    defaultInputFieldRequiredness: true,
    withInput: {
        argOptions: {
            required: true,
        },
    },
    plugins: [PothosScopeAuthPlugin, WithInputPlugin, SimpleObjectsPlugin, ZodValidationPlugin, DrizzlePlugin],
    drizzle: {
        client: ctx => drizzle(ctx.DB),
        schema,
    },
    scopeAuth: {
        authScopes: ctx => ({
            user: !!ctx.user,
        }),
        unauthorizedError: (_parent, _context, _info, result) => new GraphQLError(result.message, ERR.UNAUTHORIZED),
    },
    zod: {
        validationError: (zodError) => {
            const details = {
                extensions: {
                    code: ERR.INVALID.extensions.code,
                    http: ERR.INVALID.extensions.http,
                    errors: zodError.errors.map(err => ({
                        path: err.path,
                        message: err.message,
                    })),
                },

            };
            return new GraphQLError("Validation error", details);
        },
    },
});

builder.queryType();
builder.mutationType();

builder.addScalarType("DateTime", DateTimeResolver);
