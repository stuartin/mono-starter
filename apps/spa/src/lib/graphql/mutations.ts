import { graphql } from "@mono/shared/graphql";

export const MutationRegister = graphql(`
    mutation Register($input: MutationRegisterInput!) {
        register(
            input: $input
        ) {
            id
            name
        }
    }    
`)