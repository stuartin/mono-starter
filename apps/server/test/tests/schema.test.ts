import { schema } from "#lib/gql/schema";
import { printSchema } from "graphql";
import { describe, expect, it } from "vitest";

describe("schema", () => {
    it("matches the expected schema", () => {
        expect(printSchema(schema)).toMatchSnapshot();
    });
});
