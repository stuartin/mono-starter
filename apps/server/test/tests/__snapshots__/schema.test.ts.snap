// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schema > generates expected schema 1`] = `
""""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

type Mutation {
  login(input: MutationLoginInput!): User
  logout: Boolean
  register(input: MutationRegisterInput!): User
}

input MutationLoginInput {
  email: String!
  password: String!
}

input MutationRegisterInput {
  email: String!
  name: String!
  password: String!
}

type Query {
  me: User
}

"""User"""
type User {
  createdAt: DateTime
  id: ID
  name: String
  updatedAt: DateTime
}"
`;

exports[`schema > matches the expected schema 1`] = `
""""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

type Mutation {
  login(input: MutationLoginInput!): User
  logout: Boolean
  register(input: MutationRegisterInput!): User
}

input MutationLoginInput {
  email: String!
  password: String!
}

input MutationRegisterInput {
  email: String!
  name: String!
  password: String!
}

type Query {
  me: User
}

"""User"""
type User {
  createdAt: DateTime
  id: ID
  name: String
  updatedAt: DateTime
}"
`;
